onnx/reference/ops/op_cast.py

```diff
@@ -74,6 +77,25 @@ def cast_to(x, to, saturate):  # noqa: PLR0911
             y[i] = el
         return y.reshape(x.shape)
 
+    i4 = [
+        (uint4, "uint4", TensorProto.UINT4, False),
+        (int4, "int4", TensorProto.INT4, True),
+    ]
+    for np_type, np_desc, tensor_type, signed in i4:
+        if x.dtype == np_type and x.dtype.descr[0][0] == np_desc:
+            if to == tensor_type:
+                return x
+            to_type = tensor_dtype_to_np_dtype(to)
+            return x.astype(to_type)
+
+        if to == tensor_type:
+            xf = x.astype(np.float32).ravel()
+            y = np.empty(xf.shape, dtype=np_type).ravel()
+            for i in range(y.shape[0]):
+                el = float32_to_4bit_unpacked(xf[i], signed=signed)
```

Comment:
naming: "el": prefer more meaningful and readable variable names, here and everywhere else. Avoid abbreviations.